File: java_src/src/main/java/com/google/crypto/tink/jwt/JwtHmacKey.java
// Copyright 2023 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
////////////////////////////////////////////////////////////////////////////////

package com.google.crypto.tink.jwt;

import com.google.crypto.tink.AccessesPartialKey;
import com.google.crypto.tink.Key;
import com.google.crypto.tink.subtle.Base64;
import com.google.crypto.tink.util.SecretBytes;
import com.google.errorprone.annotations.CanIgnoreReturnValue;
import com.google.errorprone.annotations.RestrictedApi;
import java.nio.ByteBuffer;
import java.security.GeneralSecurityException;
import java.util.Optional;
import javax.annotation.Nullable;

/** Represents a JWT HMAC key to create and verify JWT using HMAC. */
class JwtHmacKey extends JwtMacKey {
  private final JwtHmacParameters parameters;
  private final SecretBytes key;
  private final Optional<Integer> idRequirement;
  private final Optional<String> kid;

  /** Helps creating new {@code JwtHmacKey} objects. */
  public static class Builder {
    private Optional<JwtHmacParameters> parameters = Optional.empty();
    private Optional<SecretBytes> keyBytes = Optional.empty();
    private Optional<Integer> idRequirement = Optional.empty();
    private Optional<String> customKid = Optional.empty();

    private Builder() {}

    @CanIgnoreReturnValue
    public Builder setParameters(JwtHmacParameters parameters) {
      this.parameters = Optional.of(parameters);
      return this;
    }

    @CanIgnoreReturnValue
    public Builder setKeyBytes(SecretBytes keyBytes) {
      this.keyBytes = Optional.of(keyBytes);
      return this;
    }

    @CanIgnoreReturnValue
    public Builder setIdRequirement(int idRequirement) {
      this.idRequirement = Optional.of(idRequirement);
      return this;
    }

    @CanIgnoreReturnValue
    public Builder setCustomKid(String customKid) {
      this.customKid = Optional.of(customKid);
      return this;
    }

    private Optional<String> computeKid() throws GeneralSecurityException {
      if (parameters
          .get()
          .getKidStrategy()
          .equals(JwtHmacParameters.KidStrategy.BASE64_ENCODED_KEY_ID)) {
        byte[] bigEndianKeyId = ByteBuffer.allocate(4).putInt(idRequirement.get()).array();
        if (customKid.isPresent()) {
          throw new GeneralSecurityException(
              "customKid must not be set for KidStrategy BASE64_ENCODED_KEY_ID");
        }
        return Optional.of(Base64.urlSafeEncode(bigEndianKeyId));
      }
      if (parameters.get().getKidStrategy().equals(JwtHmacParameters.KidStrategy.CUSTOM)) {
        if (!customKid.isPresent()) {
          throw new GeneralSecurityException("customKid needs to be set for KidStrategy CUSTOM");
        }
        return customKid;
      }
      if (parameters.get().getKidStrategy().equals(JwtHmacParameters.KidStrategy.IGNORED)) {
        if (customKid.isPresent()) {
          throw new GeneralSecurityException("customKid must not be set for KidStrategy IGNORED");
        }
        return Optional.empty();
      }
      throw new IllegalStateException("Unknown kid strategy");
    }

    public JwtHmacKey build() throws GeneralSecurityException {
      if (!parameters.isPresent()) {
        throw new GeneralSecurityException("Parameters are required");
      }
      if (!keyBytes.isPresent()) {
        throw new GeneralSecurityException("KeyBytes are required");
      }

      if (parameters.get().getKeySizeBytes() != keyBytes.get().size()) {
        throw new GeneralSecurityException("Key size mismatch");
      }

      if (parameters.get().hasIdRequirement() && !idRequirement.isPresent()) {
        throw new GeneralSecurityException(
            "Cannot create key without ID requirement with parameters with ID requirement");
      }

      if (!parameters.get().hasIdRequirement() && idRequirement.isPresent()) {
        throw new GeneralSecurityException(
            "Cannot create key with ID requirement with parameters without ID requirement");
      }

      return new JwtHmacKey(parameters.get(), keyBytes.get(), idRequirement, computeKid());
    }
  }

  @RestrictedApi(
      explanation = "Accessing parts of keys can produce unexpected incompatibilities, annotate the function with @AccessesPartialKey",
      link = "https://developers.google.com/tink/design/access_control#accessing_partial_keys",
      allowedOnPath = ".*Test\\.java",
      allowlistAnnotations = {AccessesPartialKey.class})
  public static Builder builder() {
    return new Builder();
  }

  private JwtHmacKey(
      JwtHmacParameters parameters,
      SecretBytes key,
      Optional<Integer> idRequirement,
      Optional<String> kid) {
    this.parameters = parameters;
    this.key = key;
    this.idRequirement = idRequirement;
    this.kid = kid;
  }

  @RestrictedApi(
      explanation = "Accessing parts of keys can produce unexpected incompatibilities, annotate the function with @AccessesPartialKey",
      link = "https://developers.google.com/tink/design/access_control#accessing_partial_keys",
      allowedOnPath = ".*Test\\.java",
      allowlistAnnotations = {AccessesPartialKey.class})
  public SecretBytes getKeyBytes() {
    return key;
  }

  /**
   * Returns the "kid" to be used for this key.
   *
   * <p>If present, this kid will be written into the {@code kid} header during {@code
   * computeMacAndEncode}. If absent, no kid will be written.
   *
   * <p>If present, and the {@code kid} header is present, the contents of the {@code kid} header
   * needs to match the return value of this function.
   */
  @Override
  public Optional<String> getKid() {
    return kid;
  }

  @Nullable
  @Override
  public Integer getIdRequirementOrNull() {
    return idRequirement.orElse(null);
  }

  @Override
  public JwtHmacParameters getParameters() {
    return parameters;
  }

  @Override
  public boolean equalsKey(Key o) {
    if (!(o instanceof JwtHmacKey)) {
      return false;
    }
    JwtHmacKey that = (JwtHmacKey) o;
    return that.parameters.equals(parameters)
        && that.key.equalsSecretBytes(key)
        && that.kid.equals(kid)
        && that.idRequirement.equals(idRequirement);
  }
}

File: cc/jwt/jwt_hmac_key.cc
// Copyright 2024 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
////////////////////////////////////////////////////////////////////////////////

#include "tink/jwt/jwt_hmac_key.h"

#include <string>

#include "absl/base/internal/endian.h"
#include "absl/status/status.h"
#include "absl/strings/escaping.h"
#include "absl/strings/string_view.h"
#include "absl/types/optional.h"
#include "tink/jwt/jwt_hmac_parameters.h"
#include "tink/key.h"
#include "tink/partial_key_access_token.h"
#include "tink/restricted_data.h"
#include "tink/util/status.h"
#include "tink/util/statusor.h"

namespace crypto {
namespace tink {

JwtHmacKey::Builder& JwtHmacKey::Builder::SetParameters(
    const JwtHmacParameters& parameters) {
  parameters_ = parameters;
  return *this;
}

JwtHmacKey::Builder& JwtHmacKey::Builder::SetKeyBytes(
    const RestrictedData& key_bytes) {
  key_bytes_ = key_bytes;
  return *this;
}

JwtHmacKey::Builder& JwtHmacKey::Builder::SetIdRequirement(int id_requirement) {
  id_requirement_ = id_requirement;
  return *this;
}

JwtHmacKey::Builder& JwtHmacKey::Builder::SetCustomKid(
    absl::string_view custom_kid) {
  custom_kid_ = custom_kid.data();
  return *this;
}

util::StatusOr<absl::optional<std::string>> JwtHmacKey::Builder::ComputeKid() {
  switch (parameters_->GetKidStrategy()) {
    case JwtHmacParameters::KidStrategy::kBase64EncodedKeyId: {
      if (custom_kid_.has_value()) {
        return util::Status(
            absl::StatusCode::kInvalidArgument,
            "Custom kid must not be set for KidStrategy::kBase64EncodedKeyId.");
      }
      std::string base64_kid;
      char buffer[4];
      absl::big_endian::Store32(buffer, *id_requirement_);
      absl::WebSafeBase64Escape(absl::string_view(buffer, 4), &base64_kid);
      return base64_kid;
    }
    case JwtHmacParameters::KidStrategy::kCustom: {
      if (!custom_kid_.has_value()) {
        return util::Status(absl::StatusCode::kInvalidArgument,
                            "Custom kid must be set for KidStrategy::kCustom.");
      }
      return custom_kid_;
    }
    case JwtHmacParameters::KidStrategy::kIgnored: {
      if (custom_kid_.has_value()) {
        return util::Status(
            absl::StatusCode::kInvalidArgument,
            "Custom kid must not be set for KidStrategy::kIgnored.");
      }
      return absl::nullopt;
    }
    default:
      // Should be unreachable if all valid kid strategies have been handled.
      return util::Status(absl::StatusCode::kFailedPrecondition,
                          "Unknown kid strategy.");
  }
}

util::StatusOr<JwtHmacKey> JwtHmacKey::Builder::Build(
    PartialKeyAccessToken token) {
  if (!parameters_.has_value()) {
    return util::Status(absl::StatusCode::kInvalidArgument,
                        "JWT HMAC parameters must be specified.");
  }
  if (!key_bytes_.has_value()) {
    return util::Status(absl::StatusCode::kInvalidArgument,
                        "JWT HMAC key bytes must be specified.");
  }
  if (parameters_->KeySizeInBytes() != key_bytes_->size()) {
    return util::Status(
        absl::StatusCode::kInvalidArgument,
        "Actual JWT HMAC key size does not match size specified in "
        "the parameters.");
  }
  if (parameters_->HasIdRequirement() && !id_requirement_.has_value()) {
    return util::Status(
        absl::StatusCode::kInvalidArgument,
        "Cannot create key without ID requirement with parameters with ID "
        "requirement");
  }
  if (!parameters_->HasIdRequirement() && id_requirement_.has_value()) {
    return util::Status(
        absl::StatusCode::kInvalidArgument,
        "Cannot create key with ID requirement with parameters without ID "
        "requirement");
  }
  util::StatusOr<absl::optional<std::string>> kid = ComputeKid();
  if (!kid.ok()) {
    return kid.status();
  }
  return JwtHmacKey(*parameters_, *key_bytes_, id_requirement_, *kid);
}

bool JwtHmacKey::operator==(const Key& other) const {
  const JwtHmacKey* that = dynamic_cast<const JwtHmacKey*>(&other);
  if (that == nullptr) {
    return false;
  }
  if (parameters_ != that->parameters_) {
    return false;
  }
  if (key_bytes_ != that->key_bytes_) {
    return false;
  }
  if (id_requirement_ != that->id_requirement_) {
    return false;
  }
  if (kid_ != that->kid_) {
    return false;
  }
  return true;
}

}  // namespace tink
}  // namespace crypto
